### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by tarsius** [magit/magit#2982]
Magit's performance isn't so good to say the least.  I would go as far as to say that mediocre and in certain situations outright bad performance currently is Magit's biggest problem.

We have made some improvements in the past, like e.g. caching calls to Git during a buffer refresh, but these improvements are to a large extend outweighed by the addition of new features.  Already there exist useful features that are disabled by default because they can be to costly in some cases.

One major difference between Git and Magit is that the latter shows all kinds of useful information up-front.  The major disadvantage of Magit's approach is that this information has to be always up-to-date (else it couldn't be relied on), and that doing so costs time.

To achieve good performance the cost of keeping (primarily, but not only) the status buffer up-to-date has to be reduced.  And in order to do so, some fundamental changes have to be made.  Doing that will take a while because it makes it necessary to replace some core abstractions.  The good news is that improving or even replacing those abstractions will have other benefits beside improved performance.

---

**A** These are the primary causes of bad performance:

1. After the user has performed some action, the complete status buffer is refreshed.  And that is done by recreating its content from scratch.  Every time.  All of it.

2. The required information is collected by calling `git` many times.  Until recently, there was no alternative, but now Emacs has a foreign function interface and so we can start using `libgit2` now.

3. Expensive parsing (primarily diffs and logs) is done synchronously, even when we don't need the complete result immediately (because most of it is not visible until the user scrolls).

4. The results of expensive parsing are not being cached.  (The result of calls to `git` are cached, but that mitigates the effect of (2), not (1).)

5. Many sections are created by inserting the output directly into the user-visible buffer, where it is then manipulated to look and feel they way we want to.  Without changing that, it would be very hard to address (3) and (4).

---

**B** There are two primary paths for improving performance:

1. #2959 Start using `libgit2` instead of calling `git` over and over again.

2. {TODO create issue} Completely change how sections are created (using separate parsing-buffers, which likely will be preserved as a caching mechanism) and make it possible to update sections asynchronously and independently of the containing buffer.

Luckily these two paths can be tackled in parallel, because in most cases a given section is **either** created by inserting and then "washing" `git` output (A5) **or** by processing a list of items (or a single value) using Elisp and then inserting that as the sections content (A2).

So aside from this overview, these two changes will be discussed independently, in dedicated issues.

**Comment by jkdufair** [comment]
I'd love to help contribute on #2. I've developed some decent familiarity with the section management functions. I'll keep an eye out for the new issue.

**Comment by xmo-odoo** [comment]
Just want to comment that this would be *really* cool, especially number 2: I regularly do source-rewriting on a fairly large project with rewriters which are a bit hit-or-miss (they'll false-positive and rewrite stuff they should not and the serialisation will rewrite slightly more than it should) so there needs to be a post-op filtering to create the actual commits (remove stuff or split changes across multiple commits)

add -p/checkout -p kinda work but the lack of context and the keyboard answers means it's easy to "go automatic" and end up e.g. reverting a change which should not have or staging one which shouldn't be.

Magit's diff-committed and diff-uncommitted are absolute game-changers there, they're absolutely *stellar* in UI and UX. Their issue however is that staging changes are slow, and the slowness becomes *huge* as the diff size grows, with a ~15kloc diff (wasn't kidding about the bigness) it takes about 1mn30 between me hitting `s`tage and Emacs being interactive again.

**Comment by braham-snyder** [comment]
@xmo-odoo Have you tried `magit-ediff-stage`? Although it's not quite the same as [un-]staging directly from magit's status buffer,it avoids the regeneration of the status buffer per-stage (so will likely be far faster per-stage), while retaining better interactivity than stock git.

**Comment by xmo-odoo** [comment]
@braham-snyder I'm really not fond of ediff (though that may be because I've almost never used it so I find it odd), and the complete overview of the working copy is probably the thing I love most about magit's diff views.

**Comment by Alexander-Shukaev** [comment]
One more observation.  `magit-refresh` can turn out to be very time consuming, especially for `*-refine-*`-related stuff.  Oftentimes, I have `magit-status` (or any other `magit` buffer) buffer not displayed (no window displaying any of them), but when I do changes to the code in a corresponding repository, `magit-refresh` is still invoked.  How about teaching `magit` to detect whether buffer(s) which it wants to refresh are not displayed, and thus queue actual refreshing on each of them (via e.g. `buffer-local` facilities) until they are actually displayed?

**Comment by tarsius** [comment]
I am planning to do that too. Added it to the description above.

**Comment by jypma** [comment]
There may be some good pointers [over at gitstatus](https://github.com/romkatv/gitstatus) that can help also make `magit` faster. Although someone probably ought to add the multi-threading and caching improvements of `gitstatusd` to git itself.

**Comment by pooyataher** [comment]
It seems that everybody here is aware of these issues, but just for the record I must mention that:

- magit-status takes 30 seconds to run on my fast machine.

I installed magit yesterday and I was excited about it, but for now I switch back to git at command line because `git status` responds promptly for my small projects.

**Comment by jdtsmith** [comment]
> There are _three primary paths_ for improving performance: 
> _1.  ...
> 2.
> 3.
> 4.
> 5._
> Luckily these _two paths_ can be tackled in parallel...

Looks like this one is on an expanding front :).

**Comment by plusoptix** [comment]
These performance issue is still really annoying. Please fix this! 
At gentoo the support for git.el will be removed sooner or later. So I really wanted to use magit, but it's absolutely unusable: stashing, staging, updating the view takes 20-40 seconds - on a Core i7 machine with 8 cores! 
I really want to use emacs, but so it is not possible. Please - fix it! I will sponsor chocolate.

**Comment by tarsius** [comment]
> 20-40 seconds

That's not the usual experience. Either you are using a humongous repository or something is broken for you that isn't broken for other people.

**Comment by pooyataher** [comment]
> > 20-40 seconds
> 
> That's not the usual experience. Either you are using a humongous repository or something is broken for you that isn't broken for other people.

@tarsius Thanks for getting back to me.  Whatever the issue was, luckily I cannot replicate that situation anymore!  I am now using Magit everyday.  Thank you!

**Comment by plusoptix** [comment]
OK, thank you for the feedback: I digged deeper: This happens, when you have gettext- / *.po files in your tree. If you dismiss all changes on these kind of files (we have about 20 in our repo), magit responses quite fast. 

I tried also to start the emacs without init.el, so no special hooks are activated to open gettext files.  

Is this a new bug or should I only add it here as hint for other users, which have the same issue.

**Comment by tarsius** [comment]
Actually lets look at this a bit. 20 untracked (or ignored?) files isn't much at all.

So if you have those ~20 files, then it takes 20-40 seconds to do a simple refresh (<kbd>C-g</kbd>)? Please time it. And without them it takes less than a second?

Are these files being ignore by Git (i.e. is there such a rule in `.gitignore`)? If there wasn't such a rule, does adding one make the issue go away? Even if it does, if you just added such a rule now, then remove it again for the subsequent tests.

Does the Magit status buffer show any information about these files? (Is there a `Ignored files` or `Untracked files` section?) Does removing all `magit-insert-*-files` sections from `magit-status-section-hook` fix the issue? If so, then figure out which of these functions is the cause.

**Comment by plusoptix** [comment]
More input to this issue: 
There are 20 git controller po-files in my repo
(check with $> git status | grep \\.po) 
        changed:       cs.po
        changed:       da.po
        changed:       de.po
        changed:       el.po
        changed:       en.po
        changed:       es.po
        changed:       fr.po
        changed:       hu.po
        changed:       it.po
        changed:       ja.po
        changed:       lv.po
        changed:       pl.po
        changed:       poqt_gui.pot
        changed:       pt.po
        changed:       ro.po
        changed:       ru.po
        changed:       tr.po
        changed:       uk.po
        changed:       zh_CN.po
        changed:       zh_TW.po
-> entering "g" at magit buffer
-> utimer -s
Elapsed Time: 0 days 00:00:39.198 (39.198 seconds)

$>  git checkout cs.po da.po de.po el.po en.po es.po fr.po hu.po it.po ja.po
-> revert the first half of the files
        changed:       lv.po
        changed:       pl.po
        changed:       poqt_gui.pot
        changed:       pt.po
        changed:       ro.po
        changed:       ru.po
        changed:       tr.po
        changed:       uk.po
        changed:       zh_CN.po
        changed:       zh_TW.po
-> entering "g" at magit buffer
-> utimer -s
Elapsed Time: 0 days 00:00:23.259 (23.259 seconds)

$> git checkout da.po...
->    checkout the first half, so the second half of the files stay changed:      
        changed:       lv.po
        changed:       pl.po
        changed:       poqt_gui.pot
        changed:       pt.po
        changed:       ro.po
        changed:       ru.po
        changed:       tr.po
        changed:       uk.po
        changed:       zh_CN.po
        changed:       zh_TW.po

$> git checkout *.po
 - no changed po-files left over
-> enter "g" at magit buffer
-> utimer -s
Elapsed Time: 0 days 00:00:00.158 (0.158 seconds) 

How large are our files: 
$> wc *.po
   5643   24040  189701 cs.po
   4661   17088  144576 da.po
   5428   23581  186507 de.po
   4691   19496  182921 el.po
   5623   24424  183887 en.po
   5831   26770  198447 es.po
   5872   26756  200695 fr.po
   4341   15857  135002 hu.po
   5867   26355  199595 it.po
   5318   18889  188949 ja.po
   5559   23807  190925 lv.po
   4662   18840  155765 pl.po
   5866   26783  199455 pt.po
   5744   25187  192231 ro.po
   5864   25840  232415 ru.po
   5758   24131  192257 tr.po
   4810   17666  168804 uk.po
   5280   18832  177455 zh_CN.po
   5273   18809  177316 zh_TW.po

About 300 translations per language. A lot of translations are sentences.  
There are 7 further .po files, which are not git tracked and added into the .gitignore. These files do not interfere.

**Comment by plusoptix** [comment]
At the moment I have in my magit-status.el
(defcustom magit-status-sections-hook
  '(magit-insert-status-headers
    magit-insert-merge-log
    magit-insert-rebase-sequence
    magit-insert-am-sequence
    magit-insert-sequencer-sequence
    magit-insert-bisect-output
    magit-insert-bisect-rest
    magit-insert-bisect-log
    magit-insert-untracked-files
    magit-insert-unstaged-changes
    magit-insert-staged-changes
    magit-insert-stashes
    magit-insert-unpushed-to-pushremote
    magit-insert-unpushed-to-upstream-or-recent
    magit-insert-unpulled-from-pushremote
    magit-insert-unpulled-from-upstream)

When I remove 
    magit-insert-stashes
    magit-insert-unpushed-to-pushremote
    magit-insert-unpushed-to-upstream-or-recent
    magit-insert-unpulled-from-pushremote
    magit-insert-unpulled-from-upstream

Then I can reduce the time from 38 to 10 seconds

**Comment by tarsius** [comment]
I just remembered we recently added something to make this easier: set `magit-debug-refresh` to `t` and then post the output from `*Messages*`.

Here's what those numbers normally look like:
<details>

```
Refreshing magit...
Running magit-pre-refresh-hook...done (0.002s)
Refreshing buffer `magit: magit'...
  magit-insert-error-header                          5.28e-06
  magit-insert-diff-filter-header                    3.5779e-05
  magit-insert-head-branch-header                    0.005917256
  magit-insert-upstream-branch-header                0.009664677
  magit-insert-push-branch-header                    0.00295999
  magit-insert-tags-header                           0.029366055
  magit-insert-status-headers                        0.051382107
  magit-insert-merge-log                             0.002367351
  magit-insert-rebase-sequence                       0.000558741
  magit-insert-am-sequence                           0.000267262
  magit-insert-sequencer-sequence                    0.000472803
  magit-insert-bisect-output                         0.000263354
  magit-insert-bisect-rest                           5.3672e-05
  magit-insert-bisect-log                            5.2426e-05
  magit-insert-untracked-files                       0.002597223
  magit-insert-unstaged-changes                      0.003391459
  magit-insert-staged-changes                        0.005237672
  magit-insert-stashes                               0.020132359
  magit-insert-modules                               0.002975827
  magit-insert-worktrees                             0.002223353
  magit-insert-unpushed-to-pushremote                0.041945618
  magit-insert-unpushed-to-upstream-or-recent        0.032749889
  magit-insert-unpulled-from-pushremote              0.000270675
  magit-insert-unpulled-from-upstream                0.021647526
  forge-insert-pullreqs                              0.021850482
  forge-insert-issues                                0.015691664
Refreshing buffer `magit: magit'...done (0.236s)
Running magit-post-refresh-hook...done (0.008s)
Refreshing magit...done (0.253s, cached 84/116)
```
</details>

**Comment by aaronjensen** [comment]
FYI I don't see `magit-debug-refresh` in the MELPA version.

FWIW, refresh takes about 400ms in my repo and I don't ahve forge enabled. Though I just disabled magit-todos and it brought it down to 200ms... yuck.

**Comment by kyleam** [comment]
> FYI I don't see `magit-debug-refresh` in the MELPA version.

I think that was meant to be `magit-refresh-verbose`.

**Comment by aaronjensen** [comment]
Ah, thanks. Looks like forge still takes a little time even if the repo isn't forge enabled yet, and magit-todos (specifically `magit-todos-branch-list`) is very expensive.

```
Refreshing buffer ‘magit: iheartjane’...
  magit-insert-error-header                          2e-06
  magit-insert-diff-filter-header                    6e-06
  magit-insert-head-branch-header                    0.015213
  magit-insert-upstream-branch-header                0.020369
  magit-insert-push-branch-header                    0.004845
  magit-insert-tags-header                           0.011144
  magit-insert-status-headers                        0.057213
  magit-insert-merge-log                             0.004333
  magit-insert-rebase-sequence                       0.000206
  magit-insert-am-sequence                           9.9e-05
  magit-insert-sequencer-sequence                    0.000195
  magit-insert-bisect-output                         0.000101
  magit-insert-bisect-rest                           2.6e-05
  magit-insert-bisect-log                            2.6e-05
  magit-insert-untracked-files                       0.02154
  magit-insert-unstaged-changes                      0.018993
  magit-insert-staged-changes                        0.010953
  magit-todos--insert-todos                          0.158363
  magit-insert-stashes                               0.006123
  magit-insert-unpushed-to-pushremote                0.026288
  magit-insert-unpushed-to-upstream                  0.011508000000000001
  magit-insert-unpulled-from-pushremote              8.9e-05
  magit-insert-unpulled-from-upstream                0.006936
  forge-insert-pullreqs                              0.042196
  forge-insert-issues                                5e-06
Refreshing buffer ‘magit: iheartjane’...done (0.387s)
```

**Comment by Alexander-Shukaev** [comment]
Repositories with large amount of both files and changes per commit, e.g. produced by generated code (e.g. when the source code is regenerated after an update to the code generator), make `magit-status` totally unusable.  I know me and others already complained about it many times and suggested a number of improvements.  However, I have a feeling that implementing the comprehensive list of improvements from the top might take a lot of time.  It's also not clear what's the current status of this sub-project.

Long story short, not sure if the following idea is implied by the comprehensive list of improvements at the top but I'd still mention it.  I believe a drastic performance improvement for `magit-status` could already be achieved if all the sections that potentially insert diffs are eliminated and only lists of the corresponding files affected per section are displayed instead.  This solves three main performance problems:

1.  Don't have to wait for diffs to be calculated and inserted when they might not even be needed at this point at all.
2.  Don't grow the buffer extra large as Emacs is (to say softly) notorious of not working well with extra large buffers.
3.  Don't have to rebuild/refresh that extra large buffer over and over again upon any update (even a ridiculously small one).

However, doing only that would spoil usability too indeed.  The idea is to remap the key that is currently expanding diffs (which are already inserted in the buffer) to a function that would instead create a separate buffer where sections of diff hunks are inserted for that particular file only and immediately focus it; upon quit, focus back the parent `magit-status` buffer.  Whatever actions/features/bindings (e.g. (un)staging, reverting, etc.) are available in the `magit-status` buffer for the sections of diff hunks can be replicated in such separate file diff dedicated buffer.

This does not sound like a lot of work and can be introduced as an optional mode to use for performance reasons.  What do you guys think?

**Comment by ymherklotz** [comment]
I have recently switched to MacOS from Linux and have had a few performance problems. First I had a problem with buffer editing being quite slow on large files, however, this was fixed by building a MacOS specific version of emacs with the Cocoa framework enabled, which I don't think I had before. However, now I am having problems with Magit as it seems to be very slow even on the smallest repository. Everything else seems to be working fine at the moment.

I seem to be having a performance issue with Magit on MacOS Catalina, it takes 0.6s to refresh a repository with one commit vs 0.033s on linux with the same repository. This increases to 1s on the main repository I work on and it therefore takes a while to perform any actions such as adding, removing files or committing.

I have read [Magit's guide on performance for MacOS](https://magit.vc/manual/magit/MacOS-Performance.html#MacOS-Performance), and have tried various versions of emacs from 26.3 to master (28.0). I have also tried to find posts about the same issue, and have tried their solutions but they have not worked either (such as turning off themes). Most recommended upgrading the emacs version from 26.1 which I already had done though. This issue persists if I launch emacs only with magit enabled and in the config file.

I believe that it has something to do with git taking a long time to launch from emacs, which sounds similar to the `fork` vs `vfork` issue from emacs 26.1. I am not sure though how to check if my emacs version is actually calling `vfork` correctly.

Is there anything else I should try to debug? For now I only notice this problem in magit which is why I am posting it here and it seems to happen when only magit is enabled in the config file.

From the commandline, git executes as quickly as on linux. Two magit versions below:

    Magit 20200307.319, Git 2.25.1, Emacs 26.3, gnu/linux
    Magit 20200307.319, Git 2.25.1, Emacs 28.0.50, darwin

For comparison, here is the output for linux and mac after running `magit-refresh` on both linux and mac (on a repository with one file and one commit and similar hardware):

### Linux

    Refreshing magit...
    Running magit-pre-refresh-hook...done (0.009s)
    Refreshing buffer ‘magit: newproject’...
      magit-insert-error-header                          1.653e-06
      magit-insert-diff-filter-header                    0.003189448
      magit-insert-head-branch-header                    0.002403502
      magit-insert-upstream-branch-header                6.1721e-05
      magit-insert-push-branch-header                    3.5632e-05
      magit-insert-tags-header                           0.002833871
      magit-insert-status-headers                        0.011557551
      magit-insert-merge-log                             0.001288038
      magit-insert-rebase-sequence                       0.00017291
      magit-insert-am-sequence                           0.000111263
      magit-insert-sequencer-sequence                    0.000126705
      magit-insert-bisect-output                         9.7657e-05
      magit-insert-bisect-rest                           2.0356e-05
      magit-insert-bisect-log                            1.3118e-05
      magit-insert-untracked-files                       0.002335742
      magit-insert-unstaged-changes                      0.002110061
      magit-insert-staged-changes                        0.004152358
      magit-insert-stashes                               0.001360473
      magit-insert-unpushed-to-pushremote                6.6282e-05
      magit-insert-unpushed-to-upstream-or-recent        0.008802725
      magit-insert-unpulled-from-pushremote              4.1933e-05
      magit-insert-unpulled-from-upstream                1.104e-05
    Refreshing buffer ‘magit: newproject’...done (0.037s)
    Running magit-post-refresh-hook...done (0.000s)
    Refreshing magit...done (0.049s, cached 50/73)

### Mac

    Refreshing magit...
    Running magit-pre-refresh-hook...done (0.039s)
    Refreshing buffer ‘magit: random’...
      magit-insert-error-header                          2e-06
      magit-insert-diff-filter-header                    0.055977
      magit-insert-head-branch-header                    0.057386
      magit-insert-upstream-branch-header                5.3e-05
      magit-insert-push-branch-header                    3.4e-05
      magit-insert-tags-header                           0.057151
      magit-insert-status-headers                        0.199779
      magit-insert-merge-log                             0.02891
      magit-insert-rebase-sequence                       0.000172
      magit-insert-am-sequence                           8.1e-05
      magit-insert-sequencer-sequence                    0.000156
      magit-insert-bisect-output                         8.3e-05
      magit-insert-bisect-rest                           2.2e-05
      magit-insert-bisect-log                            2.2e-05
      magit-insert-untracked-files                       0.033737
      magit-insert-unstaged-changes                      0.02892
      magit-insert-staged-changes                        0.056712
      magit-insert-stashes                               0.029054
      magit-insert-unpushed-to-pushremote                5e-05
      magit-insert-unpushed-to-upstream-or-recent        0.143981
      magit-insert-unpulled-from-pushremote              3.9e-05
      magit-insert-unpulled-from-upstream                1.2e-05
    Refreshing buffer ‘magit: random’...done (0.554s)
    Running magit-post-refresh-hook...done (0.000s)
    Refreshing magit...done (0.600s, cached 50/73)

**Comment by ymherklotz** [comment]
I have actually messed around with it a bit more and think I have found a solution! For some reason setting `magit-git-executable` to the absolute path of git solved the problem, even if it was pointing to the exact same git executable (Checked using `magit-debug-git-executable`).

I know some other people were reporting the same problem, so I hope that solves it for them too.

Now I get the following new performance, which is 2x slower than Linux, but inline with the speed of `git` on MacOS vs Linux.

<details>
<summary><strong>MacOS Performance</strong></summary>
<br>
<pre><code>
Refreshing buffer ‘magit: random’...
  magit-insert-error-header                          1e-06
  magit-insert-diff-filter-header                    2.9e-05
  magit-insert-head-branch-header                    0.008377
  magit-insert-upstream-branch-header                4.7e-05
  magit-insert-push-branch-header                    3.8e-05
  magit-insert-tags-header                           0.007386
  magit-insert-status-headers                        0.021236
  magit-insert-merge-log                             0.003121
  magit-insert-rebase-sequence                       0.000118
  magit-insert-am-sequence                           6.8e-05
  magit-insert-sequencer-sequence                    0.000126
  magit-insert-bisect-output                         7.4e-05
  magit-insert-bisect-rest                           2.1e-05
  magit-insert-bisect-log                            2e-05
  magit-insert-untracked-files                       0.00416
  magit-insert-unstaged-changes                      0.003453
  magit-insert-staged-changes                        0.006959
  magit-insert-stashes                               0.003041
  magit-insert-unpushed-to-pushremote                3.6e-05
  magit-insert-unpushed-to-upstream-or-recent        0.01527
  magit-insert-unpulled-from-pushremote              7e-05
  magit-insert-unpulled-from-upstream                8e-06
Refreshing buffer ‘magit: random’...done (0.066s)
</details>

**Comment by theckman** [comment]
@ymherklotz I'm not personally experiencing this, but have experienced similar things before when my shell was slow to start because of junk I had in my `.zshrc` file. That isn't happening to you is it?

**Comment by saghaulor** [comment]
@ymherklotz In your comment you mentioned that setting the absolute path to git helped significantly.
> I have actually messed around with it a bit more and think I have found a solution! For some reason setting `magit-git-executable` to the absolute path of git (`/usr/local/bin/git`) solved the problem, even if it was pointing to the exact same git executable (Checked using `magit-debug-git-executable`).
> 
> I know some other people were reporting the same problem, so I hope that solves it for them too.
> 
> Now I get the following new performance, which is 2x slower than Linux, but inline with the speed of `git` on MacOS vs Linux.
> 
> **MacOS Performance**
> 
> ```
> Refreshing buffer ‘magit: random’...
>   magit-insert-error-header                          1e-06
>   magit-insert-diff-filter-header                    2.9e-05
>   magit-insert-head-branch-header                    0.008377
>   magit-insert-upstream-branch-header                4.7e-05
>   magit-insert-push-branch-header                    3.8e-05
>   magit-insert-tags-header                           0.007386
>   magit-insert-status-headers                        0.021236
>   magit-insert-merge-log                             0.003121
>   magit-insert-rebase-sequence                       0.000118
>   magit-insert-am-sequence                           6.8e-05
>   magit-insert-sequencer-sequence                    0.000126
>   magit-insert-bisect-output                         7.4e-05
>   magit-insert-bisect-rest                           2.1e-05
>   magit-insert-bisect-log                            2e-05
>   magit-insert-untracked-files                       0.00416
>   magit-insert-unstaged-changes                      0.003453
>   magit-insert-staged-changes                        0.006959
>   magit-insert-stashes                               0.003041
>   magit-insert-unpushed-to-pushremote                3.6e-05
>   magit-insert-unpushed-to-upstream-or-recent        0.01527
>   magit-insert-unpulled-from-pushremote              7e-05
>   magit-insert-unpulled-from-upstream                8e-06
> Refreshing buffer ‘magit: random’...done (0.066s)
> ```

How exactly did you set the `magit-git-executable` path?  Do I just add 
```elisp
(setq magit-git-exectuable "/usr/local/bin/git")
```
to my emacs config?

I think my path was already set?

```
magit-git-executable: "/usr/local/bin/git" (2.26.2)
exec-path: ("/Users/stephenaghaulor/.nvm/versions/node/v12.7.0/bin/" "/Users/stephenaghaulor/.rbenv/shims/" "/usr/local/bin/" "/usr/bin/" "/bin/" "/usr/sbin/" "/sbin/" "/Library/Apple/usr/bin/" "/Users/stephenaghaulor/.local/bin/" "/Users/stephenaghaulor/.rbenv/bin/" "/Users/stephenaghaulor/src/gocode/bin/" "/Applications/Emacs.app/Contents/MacOS/bin-x86_64-10_14/" "/Applications/Emacs.app/Contents/MacOS/libexec-x86_64-10_14/" "/Applications/Emacs.app/Contents/MacOS/libexec/")
  /Users/stephenaghaulor/.nvm/versions/node/v12.7.0/bin/ (t)
  /Users/stephenaghaulor/.rbenv/shims/ (t)
  /usr/local/bin/ (t)
    /usr/local/bin/git (2.26.2)
  /usr/bin/ (t)
    /usr/bin/git (2.24.2 (Apple Git-127))
  /bin/ (t)
  /usr/sbin/ (t)
  /sbin/ (t)
  /Library/Apple/usr/bin/ (t)
  /Users/stephenaghaulor/.local/bin/ (nil)
  /Users/stephenaghaulor/.rbenv/bin/ (nil)
  /Users/stephenaghaulor/src/gocode/bin/ (nil)
  /Applications/Emacs.app/Contents/MacOS/bin-x86_64-10_14/ (t)
  /Applications/Emacs.app/Contents/MacOS/libexec-x86_64-10_14/ (t)
  /Applications/Emacs.app/Contents/MacOS/libexec/ (t)

```

I could really use some help, I'm seeing abysmal performance.

```
Refreshing buffer ‘magit: random’...
  magit-insert-error-header                          1e-06
  magit-insert-diff-filter-header                    0.005116
  magit-insert-head-branch-header                    0.033715
  magit-insert-upstream-branch-header                0.302397
  magit-insert-push-branch-header                    7.9e-05
  magit-insert-tags-header                           0.414043
  magit-insert-status-headers                        0.763833
  magit-insert-merge-log                             0.004443
  magit-insert-rebase-sequence                       0.000231
  magit-insert-am-sequence                           0.00011
  magit-insert-sequencer-sequence                    0.000217
  magit-insert-bisect-output                         0.000115
  magit-insert-bisect-rest                           3.3e-05
  magit-insert-bisect-log                            3.3e-05
  magit-insert-untracked-files                       1.42289
  magit-insert-unstaged-changes                      0.971694
  magit-insert-staged-changes                        0.030547
  magit-insert-stashes                               0.062513
  magit-insert-unpushed-to-pushremote                7.3e-05
  magit-insert-unpushed-to-upstream-or-recent        0.098215
  magit-insert-unpulled-from-pushremote              6.3e-05
  magit-insert-unpulled-from-upstream                0.07922
Refreshing buffer ‘magit: random’...done (4.544s)
```

Please help.

**Comment by ptn** [comment]
@saghaulor that's how I set that, but it didn't make a difference for me. What did help was disabling sections for the status buffer. In my case, the worst offenders were:

```
  (remove-hook 'magit-status-sections-hook 'magit-insert-unpushed-to-pushremote)
  (remove-hook 'magit-status-sections-hook 'magit-insert-unpushed-to-upstream-or-recent)
  (remove-hook 'magit-status-sections-hook 'magit-insert-unpulled-from-pushremote)
  (remove-hook 'magit-status-sections-hook 'magit-insert-unpulled-from-upstream)
```

Disabling sections is documented in [the official manual](https://magit.vc/manual/magit/Section-Hooks.html#Section-Hooks) (it's a single line at the bottom).

**Comment by cartesian-theatrics** [comment]
Any chance this will be addressed in the near term?  I concur this is by far the biggest problem with magit. I absolutely love this tool, but for large repos it is approaching not worth using unfortunately.

**Comment by aspiers** [comment]
`magit-refresh-verbose` was already mentioned in a previous comment over a year ago, but I guess I'm not the only one who missed that super helpful tip.   I just saw it mentioned in [this blog post](https://jakemccrary.com/blog/2020/11/14/speeding-up-magit/) via Sacha Chua's weekly newsletter, so hopefully that's of use to people reading here.

**Comment by seokbeomKim** [comment]
> 
> 
> `magit-refresh-verbose` was already mentioned in a previous comment over a year ago, but I guess I'm not the only one who missed that super helpful tip. I just saw it mentioned in [this blog post](https://jakemccrary.com/blog/2020/11/14/speeding-up-magit/) via Sacha Chua's weekly newsletter, so hopefully that's of use to people reading here.

Thanks for the link, it was really helpful in my case. 
Using magit was really slow on huge codes like Linux kernel. But after removing some hooks including above tips, I could get rid of time-consuming features which are unnecessary for my usages.

`
;; revision
(remove-hook 'magit-revision-sections-hook 'magit-insert-revision-headers)
(remove-hook 'magit-revision-sections-hook 'magit-insert-revision-notes)
(remove-hook 'magit-revision-sections-hook 'magit-insert-revision-diff)
(remove-hook 'magit-revision-sections-hook 'magit-insert-revision-tag)
(remove-hook 'magit-revision-sections-hook 'magit-insert-xref-buttons)

;; status
(remove-hook 'magit-status-sections-hook 'magit-insert-tags-header)
(remove-hook 'magit-status-sections-hook 'magit-insert-status-headers)
`

**Comment by Profpatsch** [comment]
```
  magit-insert-error-header                          2.145e-06 
  magit-insert-diff-filter-header                    0.007394417 
  magit-insert-head-branch-header                    0.006269013 
  magit-insert-upstream-branch-header                1.054e-05 
  magit-insert-push-branch-header                    9.621e-06 
  magit-insert-tags-header                           0.014738266 !
  magit-insert-status-headers                        0.032404343 !!
  magit-insert-merge-log                             0.002848184 
  magit-insert-rebase-sequence                       2.420365681 !!
  magit-insert-am-sequence                           0.000166788 
  magit-insert-sequencer-sequence                    0.000218156 
  magit-insert-bisect-output                         0.000103943 
  magit-insert-bisect-rest                           2.768e-05 
  magit-insert-bisect-log                            4.8204e-05 
  magit-insert-untracked-files                       0.014268144 !
  magit-insert-unstaged-changes                      0.00530372 
  magit-insert-staged-changes                        0.014727778 !
  magit-insert-stashes                               0.073646006 !!
  magit-insert-unpushed-to-pushremote                5.2532e-05 
  magit-insert-unpushed-to-upstream-or-recent        0.019587845 !
  magit-insert-unpulled-from-pushremote              3.435e-06 
  magit-insert-unpulled-from-upstream                3.128e-06 
```

magit-insert-rebase-sequence is taking **2.5 seconds** on a small-ish repo here with 2500 commits and 100 stashes

**Comment by jdelStrother** [comment]
@Profpatsch do you have a rebase in progress?  My magit-insert-rebase-sequence is 0.0002 seconds, on a repo with 41,307 commits and 100 stashes.

```
Refreshing buffer ‘magit: web’...
  magit-insert-error-header                          3e-06 
  magit-insert-diff-filter-header                    0.018225 !
  magit-insert-head-branch-header                    0.022502 !
  magit-insert-upstream-branch-header                0.023856 !
  magit-insert-push-branch-header                    0.009783 
  magit-insert-status-headers                        0.086622 !!
  magit-insert-merge-log                             0.009021 
  magit-insert-rebase-sequence                       0.000208 
  magit-insert-am-sequence                           9.6e-05 
  magit-insert-sequencer-sequence                    0.000189 
  magit-insert-bisect-output                         9.9e-05 
  magit-insert-bisect-rest                           3.1e-05 
  magit-insert-bisect-log                            2.9e-05 
  magit-insert-untracked-files                       0.047831 !!
  magit-insert-unstaged-changes                      0.013027 !
  magit-insert-staged-changes                        0.021695 !
  magit-insert-stashes                               0.091677 !!
  magit-insert-unpushed-to-pushremote                7.2e-05 
  magit-insert-unpushed-to-upstream-or-recent        0.131429 !!
  magit-insert-unpulled-from-pushremote              5e-06 
  magit-insert-unpulled-from-upstream                0.092876 !!
  forge-insert-pullreqs                              0.028664 !
  forge-insert-issues                                0.001083 
Refreshing buffer ‘magit: web’...done (0.553s)
Running magit-post-refresh-hook...done (0.014s)
Refreshing magit...done (0.570s, cached 56/86 (65%))
```

**Comment by Profpatsch** [comment]
@jdelStrother yes, of course, the function is a no-op without a rebase.

**Comment by Profpatsch** [comment]
it’s not a big rebase either, I’m rebasing maybe 10 commits.

**Comment by Profpatsch** [comment]
My initial profiling has pointed to https://github.com/magit/magit/blob/21454777281247d97814ce5fb64f4afe39fab5da/lisp/magit-sequence.el#L978-L1038 as the function taking most of the time.

**Comment by aspiers** [comment]
If you want to do more profiling, you may find https://github.com/aspiers/etrace useful.

**Comment by black7375** [comment]
I also think it's good to use gitstatus.
https://github.com/romkatv/gitstatus

**Comment by salutis** [comment]
> [...] bad performance currently is Magit's biggest problem.

+1 Every time I rename a directory, I use the standard command line Git to commit the change, as Magit would create a huge, 500MB or even 1GB, status buffer with diffs that nobody needs, blocking the entire Emacs environment. It would be _amazingly_ helpful if we could hit `C-g` to stop Magit from diffing for the status buffer, like we can do when committing large changes.

**Comment by tarsius** [comment]
> I have noticed similar delay and delay workarounds for other commands like committing

Lately I have noticed similar failures to redisplay for many things unrelated to Magit. I suspect there is a regression in Emacs' `emacs-28` branch concerning redisplay. Were you able to reproduce this issue using Emacs v27.1?

**Comment by tarsius** [comment]
The described symptoms are very similar, but it appears this is a different issue. I have no idea what could be wrong here.

**Comment by geza-herman** [comment]
Let me share a change in magit which made `magit-status` a little bit faster for me:
```emacs-lisp
  (defun magit-rev-format (format &optional rev args)
    (let ((str (magit-git-string "log" "-1" "--no-patch"
                                 (concat "--format=" format) args
                                 (if rev (concat rev "^{commit}") "HEAD") "--")))
      (unless (string-equal str "")
        str)))
```
In other words, use `git log -1` instead of `git show`. I'm not sure whether this always faster or not, but on the repository I'm working on, `log -1` is almost 10 times faster (the speedup can be git version dependent). Of course, `magit-rev-format` doesn't take too much time relative to the whole `magit-status`, but for me, every ms counts: I optimized the repository by removing unnecessary refs (because `--decorate` can take a lot of time, if there are a lot of refs), so `magit-status` takes only 0.2 sec. In this case, this little `log -1` optimization has a noticeable effect.

**Comment by hrehfeld** [comment]
Not sure if useful, but here is a case where `magit-git-wash` is slow while staging -- input is sluggish. the magit status buffer is 4300 lines long if expanded:

```
       12617  98% - command-execute
       12307  96%  - funcall-interactively
       12304  96%   - magit-stage
       12304  96%    - magit-apply-hunk
       12304  96%     - magit-apply-patch
       12262  95%      - magit-refresh
       12214  95%       - magit-refresh-buffer
       12204  95%        - magit-status-refresh-buffer
       12114  94%         - magit-run-section-hook
        9048  70%          - magit-insert-unstaged-changes
        9022  70%           - magit--insert-diff
        9022  70%            - magit-git-wash
        8894  69%             - magit-diff-wash-diffs
        8894  69%              - magit-wash-sequence
        8894  69%               - #<compiled -0xae28e6034cd057f>
        8894  69%                - apply
        8890  69%                 - magit-diff-wash-diff
        4851  37%                  - magit-diff-insert-file-section
        4174  32%                   - magit-wash-sequence
        2400  18%                    - magit-diff-wash-hunk
        2316  18%                     - magit-delete-line
        2316  18%                      - delete-region
        2316  18%                       - apply
        2313  18%                        - ad-Advice-delete-region
         604   4%                         + #<compiled 0x8c5ddffe3ffe>
           7   0%                           vhl/.push-to-after-change-hook
           3   0%                           #<subr delete-region>
          37   0%                     + run-hook-with-args-until-success
           6   0%                     + magit-hunk-section
           3   0%                     + mapcar
           3   0%                     + magit-insert-heading
          24   0%                   + magit-insert-heading
          18   0%                     run-hook-with-args-until-success
          11   0%                   + magit-file-section
           4   0%                     eieio-oref
        2537  19%                  - magit-delete-match
        2537  19%                   - delete-region
        2537  19%                    - apply
        2533  19%                     - ad-Advice-delete-region
         559   4%                      - #<compiled 0x8c5ddffe3ffe>
         559   4%                         vhl/.pop-from-after-change-hook
           6   0%                        vhl/.push-to-after-change-hook
         835   6%                  + magit-delete-line
           3   0%                  + cl-member-if
         116   0%             + magit-git-insert
        2102  16%          + magit-insert-unpushed-to-upstream-or-recent
         457   3%          + magit-insert-staged-changes
         187   1%          + magit-insert-stashes
         116   0%          + magit-insert-status-headers
          32   0%          + magit-insert-untracked-files
          27   0%          + magit-insert-unpulled-from-upstream
           2   0%          + magit-insert-sequencer-sequence
          86   0%         + magit-section-show
           4   0%         + magit-git-exit-code
           7   0%        + magit-section-goto-successor
          37   0%       + magit-auto-revert-buffers
          42   0%      + magit-run-git-with-input
         310   2%  + byte-code
         183   1% + timer-event-handler
```

**Comment by ackerleytng** [comment]
If we use libgit2, can magit still work over tramp? I wonder if there's value in some kind of git server that magit can connect to, and perhaps the git server can watch key files indicating git's state (and update accordingly)

**Comment by kenkangxgwe** [comment]
Can we have a lite version of the status page? 

If you work in [AOSP](https://android.googlesource.com/platform/frameworks/base/) or [chromium](https://chromium.googlesource.com/chromium/src.git), you will find the status pages are really slow to load, even though `git status` and `git log --oneline` are pretty fast in the command line.

**Comment by eeshugerman** [comment]
`git maintenance run --task gc` makes a big difference for me: `magit-status` was taking about ~3.5 seconds, now it's down to ~1.0 second.

**Comment by Atemu** [comment]
Interesting, that was the case for me aswell. Can you instruct git to schedule gcs more often somehow?

**Comment by corngood** [comment]
I'm doing quite a large rebase, where `magit-insert-rebase-sequence` takes most of the time (~5s):

```
         630  54%       - magit-insert-rebase-sequence
         625  54%        - magit-rebase-insert-merge-sequence
         342  29%         - magit-sequence-insert-sequence
         179  15%          - magit-sequence-insert-commit
         140  12%           - magit-format-rev-summary
         140  12%            - magit-rev-format
         138  12%             - magit-git-string
         128  11%              - apply
         128  11%               - magit-git-insert
         128  11%                - magit-process-git
         127  11%                 - magit-process-file
         127  11%                    process-file
           6   0%              + #<compiled -0x1ddfd105af7b8ad7>
         101   8%          - magit-rev-equal
         101   8%           - magit-git-success
         101   8%            - magit-git-exit-code
         101   8%             - magit-process-git
         100   8%              - magit-process-file
          99   8%                 process-file
           1   0%                 magit-process-environment
          55   4%          + magit-patch-id
           2   0%          + magit-git-lines
           1   0%          + magit-rev-parse
           1   0%          + magit-anything-unmerged-p
         277  24%         - magit-sequence-insert-commit
         226  19%          - magit-format-rev-summary
         225  19%           - magit-rev-format
         219  19%            - magit-git-string
         194  16%             - apply
         194  16%              - magit-git-insert
         194  16%               - magit-process-git
         193  16%                - magit-process-file
         193  16%                   process-file
          24   2%             + #<compiled -0x1ddfd1c6feac9ad7>
           2   0%          + magit-insert-heading
           1   0%          + run-hook-with-args-until-success
           1   0%          + eieio-oref
           3   0%         + magit-rebase--todo
           3   0%        + magit-rev-name
           2   0%        + magit-section-maybe-add-heading-map
```

I've removed it from `magit-status-sections-hook` for now, but I'll try to dig in to it when I get a chance.

**Comment by SteVwonder** [comment]
I was recently struggling with a slightly sluggish `magit-status`/`magit-refresh` in a large monorepo.  After turning on verbose refresh, I noticed half the time was spent in `magit-insert-untracked-files`. I enabled git's new fsmonitor daemon, and the listing of untracked files went from `0.5s` to `0.07s` and noticeably sped up `magit-refresh` from `0.983s` to `0.379s`.

Just wanted to mention the git fsmonitor here since I didn't find any other references to it in the various performance threads in this repo.  [Here is the github blog post](https://github.blog/2022-06-29-improve-git-monorepo-performance-with-a-file-system-monitor/) I followed to setup the daemon.

**Comment by tarsius** [comment]
Thanks for the hint.

Yeah, I should look into that and add a note alongside the other performance hints. (I've added that to my todo list, but it might be a while until I get around to it.)

**Comment by ibizaman** [comment]
@SteVwonder thank you! Refresh on my monorepo went from 0.740s to 0.468s. Untracked time went from 0.235s to 0.028s.

**Comment by vermiculus** [comment]
I find that `magit-insert-unpushed-to-upstream-or-recent` is particularly expensive (>7s for our monorepo!) – and I wonder if it could be cached on the hashed contents of `HEAD`, `git show-ref`, and `git config --list --local -z`, all of which take 50ms combined on macOS. It seems this section (and a couple other sections!) are focused on answering the 'where am I now? where am I going?' questions, which these would probably answer.

Partially writing this here for its relevance – partially just writing it down so I can also investigate what can be done here once I'm back from running errands.

**Comment by tarsius** [comment]
> I find that `magit-insert-unpushed-to-upstream-or-recent` is particularly expensive (>7s for our monorepo!)

That surprises me. How many commits does that normally list in that repository?

**Comment by vermiculus** [comment]
> That surprises me. How many commits does that normally list in that repository?

It's still up to the value of `magit-log-section-commit-count`, but this branch has >3.5 million commits with about 11GB of data in `.git/objects/`, so most of the time is coming from Git.

**Comment by zaphoid-beatlejuice** [comment]
Dear all,
I am an admittedly novice git and magit user, but perhaps my two cents might be of use.

`(setq magit-refresh-status-buffer nil)` did not affect my performance at all.

Working with a merge of a month's work I had a typical
magit-insert-staged-changes time of 42 seconds.

Every change (staging, unstaging, trying to select "ours" or "theirs" but only being told I could trash the file and refusing...) triggered it. The redisplay also constantly returned the cursor to the top of the buffer, which, it seemed, added insult to injury, if you follow my metaphor.

I found the idea of a lesser version of the status-buffer seductive, it would only show new/deleted/unmerged/modified + file path. I would deal with all of the files that I didn't need to look into, or use ediff. Perhaps commit the bulk of quick decisions and use a classic magit-status buffer for the rest.

All this now appears unnecessary.

Looking into code to try to bypass functions according to a flag, I found that there is already one: `(setq magit-inhibit-refresh t)` worked to effectively stop the status buffer from ever updating. (Even when pressing "g".)

Thus, after allowing the display to settle a first time, `(setq
magit-inhibit-refresh t)` allowed me to go through each file making my decisions, sometimes ediffing, without ever lossing control. And with the diff information just a tab away the whole while.

I think I might put toggling this variable into a transient, and maybe add a `(let ((magit-inhibit-refresh nil))` to `(magit-refresh)` when called with "g" to allow manual updating to override it.

I suggest this as an effective work-around for those whose workflow can deal with each file separately, or who are content to call updates manually.

**Comment by mplanchard** [comment]
I just want to report that I'm not totally sure what the cause is, and I haven't been able to really figure out anything useful from profiling, but the hooks `magit-section-pre-command-hook` and `magit-section-post-command-hook` lead to noticeably laggy input for me when working in large repositories, with the effect being significantly worse when I have many (>50) open buffers. 

My installation is essentially stock magit + forge (with magit-delta added for diffs), and this occurs with the default set of sections. The only things I've customized are turning of line-number mode via the `magit-mode-hook`, setting `forge-add-default-bindings` to `nil`, and setting `magit-save-repository-buffers` to `nil`. I am using `evil-mode`, but this happens with regular cursor movement with `evil-mode` disabled as well.

I am using the `magit-status-mode-hook` currently to call a function that removes those functions from the pre- and post- command hooks whenever I enter status mode.

Interestingly, even with these disabled, input is still laggy whenever I move my cursor into or move my cursor between lines of the header section of the status buffer, i.e. moving into or between the `Head:`, `Merge:`, `Push:`, and `Tag:` lines. This isn't as much of an issue for me since I spend most of my cursor time in the other sections, so I haven't tried to investigate it yet.

magit version: 4.0.0 5a090e8
forge version: 0.4.0 0b5571b

**Comment by nicolasdespres** [comment]
Hi everybody, thanks for creating magit. I love its UI, they way it fits in Emacs and how close it is to the command-line interface. gitk and git-gui scale well for large projects, and they use git as backend, not libgit2. So, I think magit could achieve good performance without using libgit2. Parsing asynchronously and limiting the number of updates, especially when staging/unstaging/discarding, seems like a promising improvement. Personally, I do not mind pressing 'g' manually. Maybe there is more room for improvement on the frontend side in the way magit maintains its Emacs buffers. Unfortunately, I am not a good Elisp coder to tell.

**Comment by agzam** [comment]
I'm just gonna add to this meta without reading all previous discussions, and I'm sorry - perhaps someone can point to a workaround. For whatever reasons deleting worktrees take forever. It shouldn't be like that. It blocks Emacs for a very long time whenever I try to remove a worktree.

**Comment by Ergus** [comment]
Hi:
Recently a new package [magit-prime](https://github.com/Azkae/magit-prime) pretends (is intended to) improve performance. My question is, how much is this approach intended to help? If so, could we implement something similar more "native" without requiring an external/extra package?

**Comment by tarsius** [comment]
> If so, could we implement something similar more "native" without requiring an external/extra package?

This started out as the pull-request #5360 and I suggested there that it should be maintained as a separate package for the time being.
